be/src/formats/parquet/stored_column_reader.cpp

```diff
@@ -448,35 +467,46 @@ Status OptionalStoredColumnReader::_read_values_on_levels(size_t num_values,
         size_t level_parsed = 0;
         RETURN_IF_ERROR(_decode_levels(&num_values, &level_parsed, &def_levels));
         DCHECK_EQ(num_values, level_parsed);
-        _is_nulls.resize(num_values);
+        _null_infos.reset_with_capacity(num_values);
+        size_t num_ranges = 1;
+        size_t num_nulls{};
         // decode def levels
-        for (size_t i = 0; i < num_values; ++i) {
-            _is_nulls[i] = def_levels[i] < _field->max_def_level();
-        }
-        return _reader->decode_values(num_values, &_is_nulls[0], content_type, dst, filter);
+        level_t max_def_level = _field->max_def_level();
+        uint8_t* __restrict is_nulls = _null_infos.nulls_data();
+        assign_nulls(def_levels, max_def_level, num_values, is_nulls, &num_ranges, &num_nulls);
+        _null_infos.num_ranges = num_ranges;
+        _null_infos.num_nulls = num_nulls;
+        return _reader->decode_values(num_values, _null_infos, content_type, dst, filter);
     }
 }
 
 Status RepeatedStoredColumnReader::_read_values_on_levels(size_t num_values,
                                                           starrocks::parquet::ColumnContentType content_type,
                                                           starrocks::Column* dst, bool append_default,
                                                           const FilterData* filter) {
-    _is_nulls.resize(num_values);
+    _null_infos.reset_with_capacity(num_values);
+
     int null_pos = 0;
     level_t* def_levels = _reader->def_level_decoder().get_forward_levels(num_values);
+    uint8_t* __restrict is_nulls = _null_infos.nulls_data();
+    int16_t* __restrict levels = def_levels;
+    size_t num_nulls{};
     for (int i = 0; i < num_values; ++i) {
-        level_t def_level = def_levels[i];
-        _is_nulls[null_pos] = (def_level < _field->max_def_level());
+        level_t def_level = levels[i];
+        is_nulls[null_pos] = (def_level < _field->max_def_level());
         // if current def level < ancestor def level, the ancestor will be not defined too, so that we don't
         // need to add null value to this column. Otherwise, we need to add null value to this column.
-        null_pos += (def_level >= _field->level_info.immediate_repeated_ancestor_def_level);
+        bool has_value = (def_level >= _field->level_info.immediate_repeated_ancestor_def_level);
+        num_nulls += is_nulls[null_pos] & has_value;
+        null_pos += has_value;
```

Comment:
why don't collect  num_ranges for repeated column? dose it cost too much?

Reply:
This code cannot be automatically vectorized, so num_range is not calculated for it for now.